collect_app/src/main/java/org/odk/collect/android/activities/GoogleDriveActivity.java

```diff
@@ -56,104 +59,83 @@
 import org.odk.collect.android.application.Collect;
 import org.odk.collect.android.listeners.GoogleDriveFormDownloadListener;
 import org.odk.collect.android.listeners.TaskListener;
 import org.odk.collect.android.logic.DriveListItem;
 import org.odk.collect.android.preferences.PreferenceKeys;
+import org.odk.collect.android.tasks.GoogleSheetsTask;
+import org.odk.collect.android.utilities.PlayServicesUtil;
 import org.odk.collect.android.utilities.ToastUtils;
 
-import java.io.BufferedReader;
+import java.io.ByteArrayOutputStream;
 import java.io.File;
-import java.io.FileWriter;
+import java.io.FileOutputStream;
 import java.io.IOException;
-import java.io.InputStream;
-import java.io.InputStreamReader;
 import java.util.ArrayList;
-import java.util.Collection;
 import java.util.Collections;
 import java.util.Comparator;
 import java.util.HashMap;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Stack;
 
-public class GoogleDriveActivity extends ListActivity implements GoogleApiClient.OnConnectionFailedListener,
-        TaskListener, GoogleDriveFormDownloadListener {
+import pub.devrel.easypermissions.AfterPermissionGranted;
+import pub.devrel.easypermissions.EasyPermissions;
+
+//import com.google.api.services.drive.model.ParentList;
+
+public class GoogleDriveActivity extends ListActivity implements
+        GoogleApiClient.OnConnectionFailedListener, View.OnClickListener,
+        TaskListener, GoogleDriveFormDownloadListener, EasyPermissions.PermissionCallbacks {
 
     private final static int PROGRESS_DIALOG = 1;
 
     private final static int GOOGLE_USER_DIALOG = 3;
-
+    private static final int RESOLVE_CONNECTION_REQUEST_CODE = 5555;
+    private static final int COMPLETE_AUTHORIZATION_REQUEST_CODE = 4322;
+    private static final String MY_DRIVE_KEY = "mydrive";
+    private static final String PATH_KEY = "path";
+    private static final String DRIVE_ITEMS_KEY = "drive_list";
+    private static final String PARENT_KEY = "parent";
+    private static final String ALERT_MSG_KEY = "alertmsg";
+    private static final String ALERT_SHOWING_KEY = "alertshowing";
+    private static final String ROOT_KEY = "root";
+    private static final String FILE_LIST_KEY = "fileList";
+    private static final String PARENT_ID_KEY = "parentId";
+    private static final String CURRENT_ID_KEY = "currentDir";
+    private static final String drive_fail = "Media Error: ";
+    private static final String oauth_fail = "OAUTH Error: ";
+    protected GoogleAccountCredential mCredential;
     private ProgressDialog mProgressDialog;
-
     private AlertDialog mAlertDialog;
-
     private Button mRootButton;
-
     private Button mBackButton;
-
     private Button mDownloadButton;
-
     private ImageButton mSearchButton;
-
     private EditText mSearchText;
-
-    private Stack<String> mCurrentPath = new Stack<String>();
-
+    private Stack<String> mCurrentPath = new Stack<>();
```

Comment:
Take a look at the Gradle file -- it uses Java 7 which is the recommendation for Android. Generics type inference was introduced in Java 7 - http://docs.oracle.com/javase/7/docs/technotes/guides/language/type-inference-generic-instance-creation.html